### ISSUE TITLE
Rule to detect when Multi-dollar should be used

### ISSUE THREAD

**Issue by BraisGabin** [detekt/detekt#9610]
## Expected Behavior of the rule
The idea of this rule is to promote the usage of multi-dollar strings in favor of hacks like `${"$"}` or similars

non-compilant:

```kotlin
"""${"$"}Hello"""
"""${'$'}Hello"""
val dolar = "$"
"""$dollarHello"""
```

compilant:

```kotlin
$$"""$Hello"""
```

In the third case of non-compilant we would check if any interpolation is a constant with the char `$`. I know that this definition would not catch something like this: https://github.com/detekt/detekt/pull/9196#discussion_r3761066003 but I think that it's better like this. I'm open to ideas.

## Context
I'm not sure what should we do with non-raw strings. If we should check them or not. I have mixed feelings.

```kotlin
"\$Hello" // bad
$$"$Hello" // better
// ----
$$"$Hello $$name" // bad
"\$Hello $name" // better
```

But only report when there is not any interpolation in the `String` doesn't feel good. Any thoughts?

**Comment by rastiehaiev** [comment]
Hi @BraisGabin,
I’d like to work on this issue. I propose implementing the rule with the following initial scope:

1. Run the rule only when `LanguageFeature.MultiDollarInterpolation` is enabled for the analyzed project.
2. Analyze both raw and regular strings.
3. Report interpolation expressions that are statically known to insert a single literal `$`, including direct expressions:
```kotlin
"""${"$"}Hello"""
"""${'$'}Hello"""

val a = "${"$"}Hello"
val b = "${'$'}Hello"
```
4. Also report references when analysis can prove that the referenced immutable value equals `$`, supporting both simple-name and braced interpolation:
```kotlin
val dollar = "$"

"""$dollar Hello"""
"""${dollar}Hello"""

"$dollar Hello"
"${dollar}Hello"
```

For references, I would keep the detection conservative: report only values that can be resolved statically and are guaranteed to equal `$`, such as an applicable `const val` or local immutable `val` with a constant initializer. Mutable, delegated, custom-getter, overridable, and otherwise unknown values would not be reported.

I suggest not reporting ordinary escaped dollar signs in regular strings:
```kotlin
val text = "\$Hello"
```

Choosing between `"\$Hello"` and `$$"$Hello"` seems like a style preference, particularly when a string also contains real interpolation.

Finally, I propose making the first implementation report-only, without autocorrection. A safe rewrite would need to account for every literal dollar and real interpolation in the complete string, and an incorrect transformation could change the resulting value.

@BraisGabin 
What do you think? If that's matching with what you have in mind I’d be happy to implement it.

**Comment by BraisGabin** [comment]
Your description looks really good. I agree with all of it except the points that I comment below.

>     1. Run the rule only when `LanguageFeature.MultiDollarInterpolation` is enabled for the analyzed project.

Don't, the rule should run if it is enabled, that's it.

> Finally, I propose making the first implementation report-only, without autocorrection. A safe rewrite would need to account for every literal dollar and real interpolation in the complete string, and an incorrect transformation could change the resulting value.

We don't implement autocorrection in any detekt rule so yes, I agree.

**Comment by rastiehaiev** [comment]
@BraisGabin 

Yep, that makes sense. I also noticed that detekt rules don’t currently use a check like:
```kotlin
if (!languageVersionSettings.supportsFeature(LanguageFeature.MultiDollarInterpolation)) return
```

so I won’t introduce that pattern here: the rule will simply run when it is enabled. 
And agreed regarding autocorrection; I’ll keep the initial implementation report-only. Thanks for the clarification!
